framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -949,146 +951,175 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+                    internalReceiver, null, checker.getContext());
+        } else if (node instanceof ExplicitThisLiteralNode ||
+                   node instanceof ImplicitThisLiteralNode ||
+                   node instanceof ThisLiteralNode) {
+            Receiver internalReceiver = FlowExpressions.internalReprOf(atypeFactory, node, false);
+
+            flowExprContext = new FlowExpressionContext(
+                    internalReceiver, null, checker.getContext());
         }
+
+        return flowExprContext;
     }
 
-    /**
-     * Whether to skip a contract check based on whether the condition with
-     * given expression {@code expr} is valid for the
-     * tree {@code tree} under the context {@code flowExprContext}.
-     *
-     *  @param tree The tree that is being analyzed.
-     *  @param expr The expression condition to consider.
-     *  @param flowExprContext The current context.
+    /***
+     * Returns the flow expression receiver for the {@code expression} given the
+     * {@code flowExprContext}. The expression "this" is allowed and is handled.
+     * {@code node} refers to the method invocation or variable access being analyzed.
+     * It can be used by an overriding method for special handling of expressions
+     * such as "itself" which may indicate a reference to {@code node}.
      *
-     *  @return Whether to skip the contract check.
+     * @param expression the flow expression string to be parsed.
+     * @param flowExprContext the flow expression context with respect to which the expression string is to be evaluated.
+     * @param node the Node immediately prior to which the preconditions checked by the calling method must hold true.
+     * Used by overriding implementations. Allowed to be null.
+     * @param path the TreePath from which to obtain the scope relative to which local variables are parsed.
+     * @param treeForErrorReporting the Tree used to report parsing errors via checker.report.
+     * Used by overriding implementations.
      */
-    protected boolean skipContractCheck(Tree tree, FlowExpressions.Receiver expr, FlowExpressionContext flowExprContext) {
-        // Do not add code here. Overridden versions of this method
-        // will always have a weaker condition than simply 'return false',
-        // so calling super.skipContractCheck is not useful.
+    protected FlowExpressions.Receiver parseExpressionString(String expression,
+            FlowExpressionContext flowExprContext,
+            TreePath path, Node node, Tree treeForErrorReporting) throws FlowExpressionParseException {
+        expression = expression.trim();
 
-        return false;
+        Pattern selfPattern = Pattern.compile("^(this)$");
+        Matcher selfMatcher = selfPattern.matcher(expression);
+        if (selfMatcher.matches()) {
+            expression = flowExprContext.receiver.toString().trim(); // it is possible that s == "this" after this call
```

Comment:
What `s` is the eol-comment referring to?
I would move the comment to a dedicated line.